brave/src/it/java/com/linecorp/decaton/processor/runtime/BraveTracingTest.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.runtime;
+
+import static org.junit.Assert.assertEquals;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import com.linecorp.decaton.client.DecatonClientBuilder.DefaultKafkaProducerSupplier;
+import com.linecorp.decaton.processor.TaskMetadata;
+import com.linecorp.decaton.testing.KafkaClusterRule;
+import com.linecorp.decaton.testing.TestUtils;
+import com.linecorp.decaton.testing.processor.ProcessedRecord;
+import com.linecorp.decaton.testing.processor.ProcessingGuarantee;
+import com.linecorp.decaton.testing.processor.ProcessingGuarantee.GuaranteeType;
+import com.linecorp.decaton.testing.processor.ProcessorTestSuite;
+import com.linecorp.decaton.testing.processor.ProducedRecord;
+
+import brave.Tracing;
+import brave.kafka.clients.KafkaTracing;
+
+public class BraveTracingTest {
+
+    private KafkaTracing kafkaTracing;
+    private Tracing tracing;
+
+    public static class BraveTracePropagation implements ProcessingGuarantee {
+        private final Map<String, String> producedTraceIds = new ConcurrentHashMap<>();
+        private final Map<String, String> consumedTraceIds = new ConcurrentHashMap<>();
+        private final Tracing tracing;
+
+        public BraveTracePropagation(Tracing tracing) {this.tracing = tracing;}
+
+        @Override
+        public void onProduce(ProducedRecord record) {
+            producedTraceIds.put(record.task().getId(), tracing.currentTraceContext().get().traceIdString());
```

Comment:
Yeah, with brave-tracing we have to retrieve traceId like that.
But then I think it's bit confusing with `ProducedRecord#traceId`, it's only used with `TestTracingProvider`.

Also I feel it's bit inconsistent that `ProcessorTestSuite` always embed TestTracingProvider's trace id in record header, regardless of actual tracing provider.

So how about to do like following?
- Include `Headers` in `ProducedRecord` instead of `traceId`
- In processor module's `TracingTest`, retrieve traceId from header in `onProduce`
- Add `TestTracingProducer`, which generate and embed traceId in record header upon `send` and delegate to underlying producer
  * as like simple version of brave's TracingProducer
- In processor module's `TracingTest`, install `TracingProducer` in same way as `BraveTracingTest`

In this way, ProcessorTestSuite can be agnostic from concrete tracing provider at all, and we can use TestTracingProvider or any other tracing provider for tests in consistent way

Reply:
Hmm. I worry that we're making test tracing work exactly the same way as brave tracing and will end up making assumptions that aren't valid for other tracing providers. But I don't have a better approach.